### PR TITLE
Relax supportsPrewhere check for StorageMerge

### LINKED ISSUE #61064: Prewhere over engine=Merge was broken between 24.1 and 24.2
The test:

```
DROP TABLE IF EXISTS default.test_local ON CLUSTER '{cluster}';
DROP TABLE IF EXISTS default.test_distributed ON CLUSTER '{cluster}';

CREATE TABLE default.test_local ON CLUSTER '{cluster}' 
( name String,date Date,sign Int8 ) 
ENGINE = ReplicatedCollapsingMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}', sign)
PARTITION BY date
ORDER BY name
SETTINGS index_granularity = 8192;

CREATE TABLE default.test_distributed ON CLUSTER '{cluster}' ( name String, date Date, sign Int8 ) 
ENGINE = Distributed('{cluster}', 'default', 'test_local', rand64());

SET insert_distributed_sync = 1;

INSERT INTO default.test_distributed (name, date, sign) VALUES ('1', '2024-01-01', 1),('2', '2024-01-02', 1),('3', '2024-01-03', 1),('4', '2024-01-04', 1),('5', '2024-01-05', 1),('6', '2024-01-06', 1),('7', '2024-01-07', 1),('8', '2024-01-08', 1),('9', '2024-01-09', 1),('10', '2024-01-10', 1),('11', '2024-01-11', 1);

SYSTEM SYNC REPLICA ON CLUSTER '{cluster}' default.test_local;

SELECT hostName(), * FROM clusterAllReplicas('{cluster}', 'default', 'test_local') ORDER BY date FORMAT PrettyCompactMonoBlock;

SELECT count() FROM default.test_distributed WHERE name GLOBAL IN ( SELECT name FROM default.test_distributed );

SET prefer_localhost_replica = 1;

SELECT count() FROM merge('default', '^test_distributed$') WHERE name GLOBAL IN ( SELECT name FROM default.test_distributed );
SELECT count() FROM merge('default', '^test_distributed$') PREWHERE name GLOBAL IN ( SELECT name FROM default.test_distributed );

SET prefer_localhost_replica = 0;

SELECT count() FROM merge('default', '^test_distributed$') WHERE name GLOBAL IN ( SELECT name FROM default.test_distributed );
SELECT count() FROM merge('default', '^test_distributed$') PREWHERE name GLOBAL IN ( SELECT name FROM default.test_distributed );
```

```
Received exception from server (version 24.2.1):
Code: 182. DB::Exception: Received from localhost:9000. DB::Exception: Storage Merge doesn't support PREWHERE. Stack trace:

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cf5565b
1. DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x00000000078c7743
2. DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>)::$_0::operator()(bool) const @ 0x0000000011b17056
3. DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>) @ 0x0000000011b0d305
4. DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x0000000011bce338
5. std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> std::__function::__policy_invoker<std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>::__call_impl<std::__function::__default_alloc_func<DB::registerInterpreterSelectWithUnionQuery(DB::InterpreterFactory&)::$_0, std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>>(std::__function::__policy_storage const*, DB::InterpreterFactory::Arguments const&) (.llvm.9474012881415105514) @ 0x0000000011bd4737
6. DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x0000000011af1f99
7. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000011f6616a
8. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000011f62e7a
9. DB::TCPHandler::runImpl() @ 0x0000000012fb9429
10. DB::TCPHandler::run() @ 0x0000000012fd14d9
11. Poco::Net::TCPServerConnection::start() @ 0x00000000158af152
12. Poco::Net::TCPServerDispatcher::run() @ 0x00000000158aff51
13. Poco::PooledThread::run() @ 0x00000000159a85e7
14. Poco::ThreadImpl::runnableEntry(void*) @ 0x00000000159a6c1d
15. ? @ 0x00007fa46773f609
16. ? @ 0x00007fa467664353
. (ILLEGAL_PREWHERE)
```

Discovered while testsing #59401

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Relax `supportsPrewhere` check for StorageMerge. This fixes #61064. It was hardened unnecessarily in #60082.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 7f897c382ccf12e5f5142ffb7cf9aaa9255a8eab with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61091/7f897c382ccf12e5f5142ffb7cf9aaa9255a8eab/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Storages/StorageMerge.cpp [hunk 1]
```diff
@@ -364,6 +365,18 @@ void StorageMerge::read(
     const size_t max_block_size,
     size_t num_streams)
 {
+    if (query_info.prewhere_info)
+    {
+        auto storage = getFirstTable([](const auto & table) { return !table->supportsPrewhere(); });
```
Comment (alexey-milovidov) [inline src/Storages/StorageMerge.cpp:370]: This line of code duplicates `tableSupportsPrewhere`, let's unify.

Comment (amosbird) [reply src/Storages/StorageMerge.cpp:370]: It's different. `tableSupportsPrewhere` is used to check if `optimize_move_to_prewhere` is allowed while this line of code is to check if explicit PREWHERE is valid. I'll remove `tableSupportsPrewhere` and move the implementation into `canMoveConditionsToPrewhere` instead.

**Comment by filimonov** [comment]
Any news here? @amosbird some CI/CD failures seems related to the change.

**Comment by amosbird** [comment]
I'm not sure. Will rebase to see if that resolves the issues.

**Review by KochetovNicolai — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@amosbird, this is 99% ready to merge.

Any idea, why 
> Bugfix validation — New test(s) failed to reproduce a bug 

?

**Comment by amosbird** [comment]
No "Bugfix validation — New test(s) failed to reproduce a bug" anymore.

**Comment by zcross** [comment]
> @alexey-milovidov @amosbird given it was working in previous stable builds (23.8 lts, and latest 24.1), and stopped working around 24.2, is it possible to backport the fix to 24.3/24.8 LTS versions?
> 
> Merge engine over Distributed engine is quite a frequent configuration, and prewhere is one of the key differentiator features of clickhouse that we are using in such set-ups.

+1 to this @alexey-milovidov,  would be much appreciated to have this backported into latest LTS version (like the query analyzer bugfixes)

**Pull request merged**